llama.cpp

```diff
@@ -700,32 +709,53 @@ struct llama_model_loader {
 
         if (use_mmap) {
             mapping.reset(new llama_mmap(&file_loaders.at(0)->file, prefetch_size));
-            if (!lmlock) {
-                // Don't call the callback since the actual loading will be lazy
-                // and we can't measure it.
-                progress_callback = NULL;
-            }
             if (lmlock) {
                 lmlock->init(mapping->addr);
             }
         }
 
         size_t done_size = 0;
         for (llama_load_tensor & lt : tensors_map.tensors) {
-            if (lt.ggml_tensor->backend != GGML_BACKEND_CPU) {
-                continue;
-            }
             if (progress_callback) {
                 progress_callback((float) done_size / data_size, progress_callback_user_data);
             }
             LLAMA_ASSERT(lt.ggml_tensor); // unused tensors should have been caught by load_data already
             lt.data = (uint8_t *) lt.ggml_tensor->data;
+            // allocate temp buffer if not using mmap
+            if (!use_mmap && lt.data == NULL) {
+                lt.data = (uint8_t*)malloc(ggml_nbytes(lt.ggml_tensor));
+            }
+
             load_data_for(lt);
-            lt.ggml_tensor->data = lt.data;
-            done_size += lt.size;
-            if (use_mmap && lmlock) {
-                lmlock->grow_to(done_size);
+            switch(lt.ggml_tensor->backend) {
+                case GGML_BACKEND_CPU:
+                    lt.ggml_tensor->data = lt.data;
+                    if (use_mmap && lmlock)
+                    {
+                        lock_size += lt.size;
+                        lmlock->grow_to(lock_size);
+                    }
+                    break;
+#ifdef GGML_USE_CUBLAS
+                case GGML_BACKEND_CUDA:
+                    ggml_cuda_load_data(lt.data, lt.ggml_tensor);
+                    if (!use_mmap) {
+                        free(lt.data);
+                    }
+                    break;
+#endif
+#ifdef GGML_USE_CLBLAST
+                case GGML_BACKEND_CL:
+                    ggml_cl_transform_tensor(lt.data, lt.ggml_tensor);
+                    if (!use_mmap) {
+                        free(lt.data);
```

Comment:
I think this is a good solution. Norms won't be mlocked when using OpenCL but I think that's unavoidable.

Reply:
Norms are not use CL as backend. Check the following code comments.

In order to normalize, I think it is better to support norm in OpenCL as well. I don't know how to do so though.

Reply:
I agree that the best solution would be to add an OpenCL implementation for norms. But I think it's acceptable if the norms of GPU layers are not mlocked. Compared to the weight matrices the norms are very small and can be loaded into memory relatively quickly even if the operating system doesn't have them in the file system cache.

Reply:
I see. can you guide me how to add norm for opencl? I can try to implement it.

Reply:
This is the PR where I added it: https://github.com/ggerganov/llama.cpp/pull/1483

The [mul CUDA kernel](https://github.com/JohannesGaessler/llama.cpp/blob/37f2c6c2519fe3b568978840be82bbdc93c4f089/ggml-cuda.cu#L90-L97) itself is relatively simple. For the ggml mul op I also implemented broadcasting where the values of `src1` are simply repeated across `src0` if `src1` is smaller than `src0` This can be done using the modulo operator as is done [on the CPU](https://github.com/ggerganov/llama.cpp/blob/7552ac586380f202b75b18aa216ecfefbd438d94/ggml.c#L8196-L8205) or [in CUDA](https://github.com/ggerganov/llama.cpp/blob/7552ac586380f202b75b18aa216ecfefbd438d94/ggml-cuda.cu#L536-L540C17).

For OpenCL you will probably also need to implement the boilerplate code around it that e.g. copies data from the host to the device. For CUDA I have an open PR https://github.com/ggerganov/llama.cpp/pull/1607 where I wrote a general function that manages e.g. the copying of data - but that would definitely take more work than to just write a new function analogous to one of the mul mat functions. I have never used OpenCL so you may want to ask @0cc4m for OpenCL-specific advice. Also, I have a Mumble server in case you would prefer to talk rather than write about it.

Reply:
With CUDA there was a small performance regression as well. I think the reason is that the overhead from copying data simply outweighs the time it takes to calculate the relatively small norms. Consider this plot for a different computational problem:

<details>

![pi_crude_benchmark](https://github.com/ggerganov/llama.cpp/assets/18492268/e289842c-d1f9-41fd-9b4e-4bb381293f98)

</details>

In any case, I considered fixing the memory management more important than a 1-2% performance difference.